### PR TITLE
make ReportController validate reports before saving them

### PR DESCRIPTION
**Pull request by speezepearson**
- add `verifyCanMaybe{Render,Send,Execute}` methods to various report-execution-relevant classes, a quick heuristic to check whether they could render/send/execute a given report
- make ReportController verify that a report looks executable before saving it

### CODE REVIEW

File: app/com/arpnetworking/metrics/portal/reports/Renderer.java [hunk 1]
```diff
@@ -51,4 +51,17 @@
...
+
+    /**
+     * Check whether the renderer can render a given source into a given format.
+     * Accepts all valid reports, rejects <i>some</i> invalid ones.
+     *
+     * @param source The source to render.
+     * @param format The format to render into.
+     * @throws IllegalArgumentException If the source can't be rendered into that format.
+     */
+    void verifyCanMaybeRender(
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/reports/Renderer.java]: Yeah, let's rename this.

File: app/com/arpnetworking/metrics/portal/reports/impl/EmailSender.java [hunk 2]
```diff
@@ -72,6 +73,14 @@
         });
     }
 
+    @Override
+    public void verifyCanMaybeSend(final Recipient recipient, final ImmutableCollection<ReportFormat> formatsToSend)
+            throws IllegalArgumentException {
+        if (_allowedRecipients.stream().noneMatch(pattern -> pattern.matcher(recipient.getAddress()).matches())) {
+            throw new IllegalArgumentException(recipient.getAddress() + " does not match any allowed address " + _allowedRecipients);
```
Comment (vjkoskela) [inline app/com/arpnetworking/metrics/portal/reports/impl/EmailSender.java]: Use `String.format` creating structured strings.

File: app/assets/javascripts/classes/reports/EditReportViewModel.ts [hunk 3]
```diff
@@ -110,8 +110,8 @@ class EditReportViewModel {
             dataType: "json",
             data: JSON.stringify(request),
         })
-        .fail(() => {
-            this.alertMessage(EditReportViewModel.ERROR_MESSAGE);
+        .fail((data) => {
+            this.alertMessage(`${EditReportViewModel.ERROR_MESSAGE} Code ${data.status}. Details: ${data.responseText}`);
```
Comment (vjkoskela) [inline app/assets/javascripts/classes/reports/EditReportViewModel.ts]: This should use INTL type responses to format the result for the user.

Comment (vjkoskela) [reply app/assets/javascripts/classes/reports/EditReportViewModel.ts]: Here's the PR where Brandon did this for query processing:

https://github.com/ArpNetworking/metrics-portal/pull/133

Comment (speezepearson) [reply app/assets/javascripts/classes/reports/EditReportViewModel.ts]: This is turning into a pretty extensive refactor. Fast-follow okay, to separate concerns? Other PR is already out: #292

**Pull request merged**